src/networks/competetive_net.py

```diff
@@ -238,4 +262,217 @@ def compute_loss_against(self, opponent, input):
+        state dictionary
+        """
+        return StateEncoder.encode(self.classification_layer.state_dict())
+
+    @encoded_classification_layer_parameters.setter
+    def encoded_classification_layer_parameters(self, value):
+        """
+        :param value: base64 encoded representation of the classification
+        layer's state dictionary
+        """
+        self.classification_layer.load_state_dict(StateEncoder.decode(value))
+
+    def clone(self):
+        return SSDiscriminatorNet(
+            self.loss_function,
+            self.num_classes,
+            copy.deepcopy(self.net),
+            copy.deepcopy(self.classification_layer),
+            self.data_size,
+            self.optimize_bias,
+            mnist_28x28_conv=self.mnist_28x28_conv
+        )
+
+    def _get_labeled_mask(self, batch_size):
+        label_mask = to_pytorch_variable(torch.zeros(batch_size))
+        label_count = to_pytorch_variable(torch.tensor(batch_size * self.label_rate).int())
+        label_mask[range(label_count)] = 1.0
+        return label_mask
+
+    def _log_classification_distribution(self, ground_truth, label_mask, labels,
+                                         num_usable_labels, pred_labels):
+        frequencies = labels.bincount(minlength=self.num_classes).float()
+        frequencies[frequencies == 0.0] = 1e-10
+        frequencies = frequencies.data.cpu().numpy()
+        usable_labels = label_mask * labels.float()
+
+        label_data_per_class = usable_labels.long().bincount(
+            minlength=self.num_classes
+        )
+        label_data_per_class[0] = num_usable_labels - label_data_per_class[1:].sum()
+        predicted_frequencies = np.bincount(
+            pred_labels[np.where(pred_labels == ground_truth)],
+            minlength=self.num_classes
+        )
+        classification_rate_per_class = torch.from_numpy(
+            predicted_frequencies * 100.0 / frequencies
+        )
+
+        labels = to_pytorch_variable(torch.tensor([i for i in range(0, self.num_classes)]))
+        statistics = zip(labels, label_data_per_class, classification_rate_per_class)
+
+        stats = f"\nLabel, Number of Labeled Data points, Classification Rate for this label"
+        for (label, label_data, classification_rate) in statistics:
+            stats += f"\n{label}, {label_data}, {classification_rate}"
+        self._logger.info(stats)
+
+    def compute_loss_against(self, opponent, input, labels=None, alpha=None,
+                             beta=None, iter=None, log_class_distribution=False):
+        batch_size = input.size(0)
+        tensor = torch.Tensor(batch_size)
+        tensor.fill_(self.num_classes)
+        tensor = tensor.long()
+        fake_labels = to_pytorch_variable(tensor)
+
+        label_mask = self._get_labeled_mask(batch_size)
+
+        # Positive Label Smoothing
+        real = torch.Tensor(batch_size)
+        real.fill_(0.9)
+        real = to_pytorch_variable(real)
```

Comment:
All these add-ons to the regular training should be documented. Because it has been a really long journey that you have traveled and it would be a pity if do not include all the steps in the report/paper.

Reply:
Definitely Jamal. They are currently on our SuperUROP Diary Google Doc but I'll add comments here to explain why we needed it